File: brainfuck/brainfuck.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

func main() {
	var code string
	if len(os.Args) > 1 {
		data, err := os.ReadFile(os.Args[1])
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
			os.Exit(1)
		}
		code = string(data)
	} else {
		data, err := io.ReadAll(os.Stdin)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading stdin: %v\n", err)
			os.Exit(1)
		}
		code = string(data)
	}

	tape := make([]byte, 30000)
	codePos := 0
	tapePos := 0

	for codePos < len(code) && tapePos >= 0 && tapePos < len(tape) {
		switch code[codePos] {
		case '[':
			if tape[tapePos] == 0 {
				depth := 1
				codePos++
				for depth > 0 && codePos < len(code) {
					if code[codePos] == '[' {
						depth++
					} else if code[codePos] == ']' {
						depth--
					}
					codePos++
				}
				codePos--
			} else {
				codePos++
			}
		case ']':
			if tape[tapePos] != 0 {
				depth := 1
				codePos--
				for depth > 0 && codePos >= 0 {
					if code[codePos] == ']' {
						depth++
					} else if code[codePos] == '[' {
						depth--
					}
					codePos--
				}
				codePos++
			} else {
				codePos++
			}
		case '+':
			tape[tapePos]++
			codePos++
		case '-':
			tape[tapePos]--
			codePos++
		case '>':
			tapePos++
			codePos++
		case '<':
			tapePos--
			codePos++
		case '.':
			fmt.Print(string(tape[tapePos]))
			codePos++
		case ',':
			tape[tapePos] = []byte{0}[0]
			reader := bufio.NewReader(os.Stdin)
			ch, _, _ := reader.ReadRune()
			tape[tapePos] = byte(ch)
			codePos++
		default:
			codePos++
		}
	}
}

File: brainfuck/brainfuck.nim
import os

let code = if paramCount() > 0: readFile paramStr(1)
           else: readAll stdin

var
    tape = newSeq[char](30000)
    codePos = 0
    tapePos = 0

{.push overflowchecks: off.}
proc xinc(c: var char) = inc c
proc xdec(c: var char) = dec c
{.pop.}

proc run(): bool =
  while tapePos >= 0 and tapePos < tape.len and codePos < code.len:
    case code[codePos]
    of '[':
      if tape[tapePos] == '\0':
        var depth = 1
        inc codePos
        while depth > 0 and codePos < code.len:
          if code[codePos] == '[': inc depth
          elif code[codePos] == ']': dec depth
          inc codePos
        dec codePos
      else:
        inc codePos
    of ']':
      if tape[tapePos] != '\0':
        var depth = 1
        dec codePos
        while depth > 0 and codePos >= 0:
          if code[codePos] == ']': inc depth
          elif code[codePos] == '[': dec depth
          dec codePos
        inc codePos
      else:
        inc codePos
    of '+': xinc tape[tapePos]; inc codePos
    of '-': xdec tape[tapePos]; inc codePos
    of '>': inc tapePos; inc codePos
    of '<': dec tapePos; inc codePos
    of '.': stdout.write tape[tapePos]; inc codePos
    of ',': tape[tapePos] = stdin.readChar; inc codePos
    else: inc codePos

  return true

discard run()
